### PR TITLE
Implement index case markers

### LINKED ISSUE #143: Implement the Index case on Map view
Implement Index case on the Map view

image

A structure is tagged as an index case by having a property `is_index_case = true` 

- [ ] Draw a boundary around the index with a radius of 1km.

- [ ] The radius should be configurable.

- [ ] Display a target around a structure that is identified as the index case. The target should be visible at higher zoom levels.

The target is a symbol layer with the below image 

icon-index-map

### PR DESCRIPTION
**Pull request by vincent-karuri**
Closes #143

### CODE REVIEW

File: opensrp-reveal/build.gradle [hunk 1]
```diff
@@ -204,7 +204,7 @@ dependencies {
         exclude group: 'com.android.support', module: 'appcompat-v7'
     }
 
-    implementation('org.smartregister:opensrp-client-native-form:1.5.6-SNAPSHOT@aar') {
+    implementation('org.smartregister:opensrp-client-native-form:1.5.11-PREVIEW-SNAPSHOT@aar') {
```
Comment (githengi) [inline opensrp-reveal/build.gradle]: Remove this Since the PR does not have any JSON form dependencies

File: opensrp-reveal/src/main/java/org/smartregister/reveal/util/RevealMapHelper.java [hunk 2]
```diff
@@ -53,16 +86,85 @@ public static void addSymbolLayers(@NonNull Style mMapboxMapStyle, Context conte
...
+                Location centre = (new RevealMappingHelper()).getCenter(indexCase.geometry().toJson());
+                JSONObject feature = new JSONObject(indexCase.toJson());
+                JSONObject geometry = new JSONObject();
+                geometry.put("type", "Point");
+                geometry.put("coordinates", new JSONArray(new Double[]{centre.getLongitude(), centre.getLatitude()}));
+                feature.put("geometry", geometry);
+                indexCaseSource.setGeoJson(Feature.fromJson(feature.toString()));
+
+                CircleLayer circleLayer = new CircleLayer(INDEX_CASE_CIRCLE_LAYER, INDEX_CASE_SOURCE);
+                float circleRadius = calculateZoomLevelRadius(mapboxMap, centre.getLatitude(), 500);
```
Comment (githengi) [inline opensrp-reveal/src/main/java/org/smartregister/reveal/util/RevealMapHelper.java]: Remove the 500 and use 1km Radius, also  this should be configurable on server, so implement overriding via server config

File: opensrp-reveal/src/main/java/org/smartregister/reveal/view/ListTasksActivity.java [hunk 3]
```diff
@@ -213,10 +217,17 @@ public void onStyleLoaded(@NonNull Style style) {
...
-                        RevealMapHelper.addSymbolLayers(style, ListTasksActivity.this);
+                        RevealMapHelper.addCustomLayers(style, ListTasksActivity.this);
                     }
                 });
+
                 mMapboxMap = mapboxMap;
+                mMapboxMap.addOnCameraMoveListener(new MapboxMap.OnCameraMoveListener() {
+                    @Override
+                    public void onCameraMove() {
+                        addCaseIndexBoundary(mMapboxMap, getContext());
```
Comment (githengi) [inline opensrp-reveal/src/main/java/org/smartregister/reveal/view/ListTasksActivity.java]: Why is this, here why not add index case when the GeoJSON Sources changes i.e one changes the Plan or OA, since there will be only one index case per Plan and OA

File: opensrp-reveal/src/main/java/org/smartregister/reveal/view/ListTasksActivity.java [hunk 4]
```diff
@@ -308,16 +319,29 @@ public void setGeoJsonSource(@NonNull FeatureCollection featureCollection, Featu
             geoJsonSource.setGeoJson(featureCollection);
             if (operationalArea != null) {
                 CameraPosition cameraPosition = mMapboxMap.getCameraForGeometry(operationalArea.geometry());
+                if (getInterventionLabel() == R.string.focus_investigation) {
+                    Feature indexCase = getIndexCase(mMapboxMap.getStyle().getSourceAs(getContext().getString(R.string.reveal_datasource_name)));
+                    if (indexCase != null) {
+                        cameraPosition = mMapboxMap.getCameraForGeometry(indexCase.geometry());
```
Comment (githengi) [inline opensrp-reveal/src/main/java/org/smartregister/reveal/view/ListTasksActivity.java]: Use  CameraPosition Builder to construct a camera position with target `indexCase` cordinates and specify a zoom. Level. This method would zoom to very low zoom level where the map is not clearly visible and would force user to zoom out

File: opensrp-reveal/src/main/java/org/smartregister/reveal/view/ListTasksActivity.java [hunk 5]
```diff
@@ -308,16 +319,29 @@ public void setGeoJsonSource(@NonNull FeatureCollection featureCollection, Featu
...
+                    }
+                } else {
+                    removeCaseIndexBoundary(mMapboxMap.getStyle());
```
Comment (githengi) [inline opensrp-reveal/src/main/java/org/smartregister/reveal/view/ListTasksActivity.java]: I this should be called only if one is not in a focus investigation and the the layer was already added. 
If one is in a focus investigation the GeoJSON source should be modified like the OA boundary layer

File: opensrp-reveal/src/main/java/org/smartregister/reveal/view/ListTasksActivity.java [hunk 6]
```diff
@@ -308,16 +319,29 @@ public void setGeoJsonSource(@NonNull FeatureCollection featureCollection, Featu
...
                 if (boundaryLayer == null) {
                     boundaryLayer = createBoundaryLayer(operationalArea);
                     kujakuMapView.addLayer(boundaryLayer);
-
                 } else {
                     boundaryLayer.updateFeatures(FeatureCollection.fromFeature(operationalArea));
                 }
+
+                if (getInterventionLabel() == R.string.focus_investigation) {
+                    addCaseIndexBoundary(mMapboxMap, getContext());
```
Comment (githengi) [inline opensrp-reveal/src/main/java/org/smartregister/reveal/view/ListTasksActivity.java]: Maybe you should use the above example for the boundary layer. Create layer is not already created, otherwise just update the GeoJSON Source

File: opensrp-reveal/src/main/java/org/smartregister/reveal/view/ListTasksActivity.java [hunk 7]
```diff
@@ -486,6 +510,9 @@ public void onSyncInProgress(FetchStatus fetchStatus) {
         } else if (fetchStatus.equals(FetchStatus.fetched)
                 || fetchStatus.equals(FetchStatus.nothingFetched)) {
             Snackbar.make(rootView, org.smartregister.R.string.sync_complete, Snackbar.LENGTH_LONG).show();
+            if (getInterventionLabel() == R.string.focus_investigation) {
+                addCaseIndexBoundary(mMapboxMap, getContext());
```
Comment (githengi) [inline opensrp-reveal/src/main/java/org/smartregister/reveal/view/ListTasksActivity.java]: This should be removed

File: opensrp-reveal/src/main/java/org/smartregister/reveal/util/RevealMapHelper.java [hunk 8]
```diff
@@ -53,16 +86,85 @@ public static void addSymbolLayers(@NonNull Style mMapboxMapStyle, Context conte
...
+                        iconImage(INDEX_CASE_TARGET_ICON),
+                        iconSize(dynamicIconSize));
+                symbolLayer.setFilter(eq(get(IS_INDEX_CASE), true));
+                if (mMapboxMapStyle.getLayer(INDEX_CASE_SYMBOL_LAYER) == null) {
+                    mMapboxMapStyle.addLayer(symbolLayer);
+                }
+            }
+
+            // index case circle layer
+            GeoJsonSource indexCaseSource = new GeoJsonSource(INDEX_CASE_SOURCE);
```
Comment (githengi) [inline opensrp-reveal/src/main/java/org/smartregister/reveal/util/RevealMapHelper.java]: This method needs to be on its own method, the lines upto 118, can be called only once in the app.

From here you can break down so that you two methods for Creating index case circle layer and another for updating it in case it already exists

**Comment by githengi** [comment]
I added a feature with `index case circle layer` and the Index case target and circle did not show up

File: opensrp-reveal/build.gradle (hunk 1 above)
Comment (vincent-karuri) [reply opensrp-reveal/build.gradle]: Upgraded.

File: opensrp-reveal/src/main/java/org/smartregister/reveal/util/RevealMapHelper.java (hunk 2 above)
Comment (vincent-karuri) [reply opensrp-reveal/src/main/java/org/smartregister/reveal/util/RevealMapHelper.java]: Done.

File: opensrp-reveal/src/main/java/org/smartregister/reveal/view/ListTasksActivity.java (hunk 3 above)
Comment (vincent-karuri) [reply opensrp-reveal/src/main/java/org/smartregister/reveal/view/ListTasksActivity.java]: This redraws with zoom.

File: opensrp-reveal/src/main/java/org/smartregister/reveal/view/ListTasksActivity.java (hunk 4 above)
Comment (vincent-karuri) [reply opensrp-reveal/src/main/java/org/smartregister/reveal/view/ListTasksActivity.java]: Done.

File: opensrp-reveal/src/main/java/org/smartregister/reveal/view/ListTasksActivity.java (hunk 7 above)
Comment (vincent-karuri) [reply opensrp-reveal/src/main/java/org/smartregister/reveal/view/ListTasksActivity.java]: Doesn't that only happen if new structures were fetched?

File: opensrp-reveal/src/main/java/org/smartregister/reveal/util/RevealMapHelper.java (hunk 8 above)
Comment (vincent-karuri) [reply opensrp-reveal/src/main/java/org/smartregister/reveal/util/RevealMapHelper.java]: Could you rephrase what the issue is here?

File: opensrp-reveal/src/main/java/org/smartregister/reveal/view/ListTasksActivity.java (hunk 6 above)
Comment (vincent-karuri) [reply opensrp-reveal/src/main/java/org/smartregister/reveal/view/ListTasksActivity.java]: Don't think I understand what the issue is here, could you rephrase?

File: opensrp-reveal/src/main/java/org/smartregister/reveal/view/ListTasksActivity.java (hunk 5 above)
Comment (vincent-karuri) [reply opensrp-reveal/src/main/java/org/smartregister/reveal/view/ListTasksActivity.java]: Don't think I understand what the issue is here, could you rephrase?

**Comment by vincent-karuri** [comment]
> I added a feature with `index case circle layer` and the Index case target and circle did not show up

Do you mean you added a feature with `is_index_case = true` and the index case markers were not drawn?

File: opensrp-reveal/src/main/java/org/smartregister/reveal/view/ListTasksActivity.java (hunk 3 above)
Comment (githengi) [reply opensrp-reveal/src/main/java/org/smartregister/reveal/view/ListTasksActivity.java]: This is rather ineffecient calling this every time one moves/tilts/zoom the map

File: opensrp-reveal/src/main/java/org/smartregister/reveal/view/ListTasksActivity.java (hunk 5 above)
Comment (githengi) [reply opensrp-reveal/src/main/java/org/smartregister/reveal/view/ListTasksActivity.java]: You should remove the index case layers only when a user chooses a Non FI plan. If its a FI plan then update the GeoJSON source instead of removing layers they re-adding them

File: opensrp-reveal/src/main/java/org/smartregister/reveal/view/ListTasksActivity.java (hunk 6 above)
Comment (githengi) [reply opensrp-reveal/src/main/java/org/smartregister/reveal/view/ListTasksActivity.java]: Create the index case layers once, then update the source just like how the operational area boundaries and labels are drawnin the above snippet.

```
            if (boundaryLayer == null) {
                    boundaryLayer = createBoundaryLayer(operationalArea);
                    kujakuMapView.addLayer(boundaryLayer);

                } else {
                    boundaryLayer.updateFeatures(FeatureCollection.fromFeature(operationalArea));
                }
```

File: opensrp-reveal/src/main/java/org/smartregister/reveal/view/ListTasksActivity.java (hunk 7 above)
Comment (githengi) [reply opensrp-reveal/src/main/java/org/smartregister/reveal/view/ListTasksActivity.java]: No, also when structures are changed e.g due to spray forms updating the business status

**Comment by githengi** [comment]
> > I added a feature with `index case circle layer` and the Index case target and circle did not show up
> 
> Do you mean you added a feature with `is_index_case = true` and the index case markers were not drawn?

Yes

File: opensrp-reveal/src/main/java/org/smartregister/reveal/util/RevealMapHelper.java (hunk 8 above)
Comment (githengi) [reply opensrp-reveal/src/main/java/org/smartregister/reveal/util/RevealMapHelper.java]: You need to separate Creating Layers from Creating the Sources. Layers should be created only once. However for the sources it be possible to create a method that can update the source.

File: opensrp-reveal/src/main/java/org/smartregister/reveal/view/ListTasksActivity.java (hunk 3 above)
Comment (vincent-karuri) [reply opensrp-reveal/src/main/java/org/smartregister/reveal/view/ListTasksActivity.java]: That's the only way I know it resizes on zoom. Additionally, this is not an expensive operation. Do you have an alternative for redrawing on zoom?

File: opensrp-reveal/src/main/java/org/smartregister/reveal/view/ListTasksActivity.java (hunk 5 above)
Comment (vincent-karuri) [reply opensrp-reveal/src/main/java/org/smartregister/reveal/view/ListTasksActivity.java]: They are not removed if it is a FI. That's what the first part of the if statement looks for?

File: opensrp-reveal/src/main/java/org/smartregister/reveal/view/ListTasksActivity.java (hunk 7 above)
Comment (vincent-karuri) [reply opensrp-reveal/src/main/java/org/smartregister/reveal/view/ListTasksActivity.java]: Done.

File: opensrp-reveal/src/main/java/org/smartregister/reveal/util/RevealMapHelper.java (hunk 8 above)
Comment (vincent-karuri) [reply opensrp-reveal/src/main/java/org/smartregister/reveal/util/RevealMapHelper.java]: Done.
